technology_and_tooling/testing/mocking.md

```diff
@@ -0,0 +1,526 @@
+
+~~~python
+import pytest
+import sqlite3
+from sqlite_example import query_database
+from unittest.mock import patch
+
+def test_query_db_mocked_connection():
+    """Mock the database connection and cursor to ensure the correct methods
+    are called within the query_database function
+    """
+    with patch('sqlite3.connect') as mock_connection:
+        mock_cursor = mock_connection.return_value.cursor.return_value
+        mock_cursor.fetchall.return_value = [("Jerry", "Mouse", 1)]
+        # make our fake connection
+        conn = mock_connection("my_non_existent_file")
+        sql = "SELECT * FROM Animals"
+        # test what query_database does with our fake connection
+        result = query_database(sql, connection=conn)
+        assert result[0] == ("Jerry", "Mouse", 1)
+        # check that query_database passes our SQL string to cursor.execute()
+        mock_cursor.execute.assert_called_once_with(sql)
+        # check that fetchall() was called
+        mock_cursor.fetchall.assert_called_once()
+        # check that query_database closes the connection
+        conn.close.assert_called_once()
+~~~
+
+:::
+::::
+
+Now lets look at using a function decorator.
+
+::::challenge{id=patch-decorator title="Using the `@patch` decorator"}
+
+Rewrite `test_query_db_mocked_connection` to use a function decorator instead of a context manager. You can view the [unittest documentation](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) as a guide.
+
+:::solution
+~~~python
+@patch('sqlite3.connect')
+def test_query_db_mocked_connection(mock_connection):
+    mock_cursor = mock_connection.return_value.cursor.return_value
+    mock_cursor.fetchall.return_value = [("Jerry", "Mouse", 1)]
+    # make our fake connection
+    conn = mock_connection("my_non_existent_file")
+    sql = "SELECT * FROM Animals"
+    # test what query_database does with our fake connection
+    result = query_database(sql, connection=conn)
+    assert result[0] == ("Jerry", "Mouse", 1)
+    # check that query_database passes our SQL string to cursor.execute()
+    mock_cursor.execute.assert_called_once_with(sql)
+    # check that fetchall() was called
+    mock_cursor.fetchall.assert_called_once()
+    # check that query_database closes the connection
+    conn.close.assert_called_once()
+~~~
+
+:::
+::::
+
+### Mocking using other libraries
+
+#### Monkeypatching in pytest
+
+As an alternative to using the `unitest.mock` library, its possible to use a version of mocking from within `pytest`, termed *monkeypatching*. A built-in fixture called `monkeypatch` allows modifying attributes, functions or classes within the scope of the test function. Some example methods are:
```

Comment:
is monkeypatching an alternative to unittest.mock.patch? You didn't really explain why/if monkeypatching is better than unittest.mock.patch, or just a different tool to use. Needs a bit more info/motivation here.